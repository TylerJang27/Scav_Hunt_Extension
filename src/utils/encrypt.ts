import { encrypt, decrypt } from "crypto-js/aes";
import Utf8 from "crypto-js/enc-utf8";

export const Encrypt = (
  text: string,
  encrypted: boolean,
  secretKey: string
) => {
  if (encrypted) {
    return encrypt(text, secretKey).toString();
  }
  return text;
};

export const Decrypt = (
  text: string,
  encrypted: boolean,
  secretKey: string
) => {
  if (encrypted) {
    const bytes = decrypt(text, secretKey);
    return bytes.toString(Utf8);
  }
  return text;
};

export const wrapDecrypt = (
  text: string | undefined,
  encoded: boolean,
  secretKey: string
) => {
  if (text) {
    return Decrypt(text, encoded, secretKey);
  }
  return undefined;
};

export const wrapEncrypt = (
  text: string | undefined,
  encoded: boolean,
  secretKey: string
) => {
  if (text) {
    return Encrypt(text, encoded, secretKey);
  }
  return undefined;
};
