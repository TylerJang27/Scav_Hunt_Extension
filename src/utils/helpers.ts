export const nonNull = (val: any) => val !== undefined && val !== null;
