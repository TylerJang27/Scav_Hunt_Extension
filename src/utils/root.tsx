import React from "react";
import { createRoot } from "react-dom/client";

const getRoot = () => document.getElementById("root");

export const Render = (element: JSX.Element) => {
  const root = createRoot(getRoot()!);
  root.render(<React.StrictMode>{element}</React.StrictMode>);
};
