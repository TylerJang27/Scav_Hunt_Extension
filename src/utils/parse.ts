import { DEFAULT_BACKGROUND } from "src/providers/helpers";
import {
  BulkError,
  ConfigError,
  InvalidIndexError,
  MissingFieldError,
  MissingValueError,
  SUPPORTED_VERSIONS,
  UnsupportedVersionError,
  XORFieldsError,
} from "src/types/errors";
import { ClueConfig, HuntConfig } from "src/types/hunt_config";
import { Decrypt, Encrypt, wrapDecrypt, wrapEncrypt } from "src/utils/encrypt";
import { nonNull } from "src/utils/helpers";

const ValidateRequiredField = (
  value: any,
  fieldName: string,
  index?: number,
): any => {
  if (!nonNull(value)) {
    throw new MissingFieldError(fieldName, index);
  }
  return value;
};

const ValidateRequiredNonEmptyField = (
  value: any,
  fieldName: string,
  index?: number,
): any => {
  if (!nonNull(value)) {
    throw new MissingFieldError(fieldName, index);
  }
  if (value === "") {
    throw new MissingValueError(fieldName, index);
  }
  return value;
};

// trunk-ignore(eslint/@typescript-eslint/no-unused-vars)
const ValidateXORFields = (
  value1: any,
  fieldName1: string,
  value2: any,
  fieldName2: string,
  index?: number,
) => {
  if ((!value1 && !value2) || (Boolean(value1) && Boolean(value2))) {
    throw new XORFieldsError(value1, fieldName1, value2, fieldName2, index);
  }
};

const ValidateVersion = (value: string): string => {
  if (!nonNull(value)) {
    throw new MissingFieldError("version");
  } else if (!SUPPORTED_VERSIONS.includes(value)) {
    throw new UnsupportedVersionError(value);
  }
  return value;
};

export const ParseClue = (
  { id, url, text, html, image, alt, interactive }: ClueConfig,
  index?: number,
): ClueConfig => {
  const ret = {
    id: ValidateRequiredNonEmptyField(id, "id", index) as number,
    url: ValidateRequiredNonEmptyField(url, "url", index) as string,
    text: ValidateRequiredNonEmptyField(text, "text", index) as string,
    html: html,
    image: image,
    alt: alt,
    interactive: interactive,
  };

  // TODO: TYLER ONCE WE ADD MARKDOWN SUPPORT, WE SHOULD USE XOR HERE AND REMOVE text VALIDATION
  // ValidateXORFields(text, "text", html, "html", index);

  // Validate clue config
  if (interactive) {
    ValidateRequiredNonEmptyField(interactive?.key, "interactive.key", index);
  }

  return ret;
};

const ValidateClues = (clues: ClueConfig[]) => {
  let index = 1;
  clues.forEach((clue) => {
    if (clue.id != index) {
      throw new InvalidIndexError(clue.id, index);
    }
    index++;
  });
};

// TODO: TYLER TEST WITH BAD TYPE INPUTS
const ParseHuntOptions = ({
  name,
  description,
  version,
  author,
  encrypted = false,
  background = DEFAULT_BACKGROUND,
  options = { silent: false },
  beginning,
}: HuntConfig): HuntConfig => ({
  name: ValidateRequiredNonEmptyField(name, "name") as string,
  description: ValidateRequiredNonEmptyField(
    description,
    "description",
  ) as string,
  version: ValidateVersion(version),
  author: ValidateRequiredField(author, "author") as string,
  encrypted: encrypted,
  background: background,
  options: options,
  beginning: ValidateRequiredNonEmptyField(beginning, "beginning") as string,
  clues: [],
});

export const ParseConfig = (object: any) => {
  // Parse clues
  // trunk-ignore(eslint/@typescript-eslint/no-unsafe-member-access)
  if (!object.clues || object.clues.length === 0) {
    throw new MissingValueError("clues");
  }

  // trunk-ignore(eslint)
  const [_, clues, errors] = object.clues.reduce(
    (
      [index, clueAccumulator, errorAccumulator]: [
        number,
        ClueConfig[],
        ConfigError[],
      ],
      clue: ClueConfig,
    ) => {
      try {
        return [
          index + 1,
          clueAccumulator.concat(ParseClue(clue, index)),
          errorAccumulator,
        ];
      } catch (error) {
        if (error instanceof ConfigError) {
          return [index + 1, clueAccumulator, errorAccumulator.concat(error)];
        }
        throw error;
      }
    },
    [0, [] as ClueConfig[], [] as ConfigError[]],
  );
  const parsedClues = clues as ClueConfig[];
  const parsedErrors = errors as ConfigError[];

  // Errors at this point are high enough priority that we throw so users can resolve them.
  if (parsedErrors.length) {
    throw new BulkError(parsedErrors);
  }

  try {
    // Parse hunt config
    const huntConfig: HuntConfig = ParseHuntOptions(object as HuntConfig);
    ValidateClues(parsedClues);
    huntConfig.clues = parsedClues;

    // Throw bulk error, config errors first
    if (parsedErrors.length) {
      throw new BulkError(parsedErrors);
    }
    return huntConfig;
  } catch (error) {
    // Throw bulk error with clue config errors
    if (error instanceof ConfigError) {
      throw new BulkError([error].concat(parsedErrors));
    }
    throw error;
  }
};

export const DecryptClue = (
  clue: ClueConfig,
  encrypted: boolean,
  secretKey: string,
) => ({
  /* Decoded fields */
  id: clue.id,
  /* Encoded fields */
  url: Decrypt(clue.url, encrypted, secretKey),
  text: wrapDecrypt(clue.text, encrypted, secretKey),
  html: wrapDecrypt(clue.html, encrypted, secretKey),
  image: wrapDecrypt(clue.image, encrypted, secretKey),
  alt: wrapDecrypt(clue.alt, encrypted, secretKey),
  interactive: clue.interactive
    ? {
        prompt: wrapDecrypt(clue.interactive.prompt, encrypted, secretKey),
        key: clue.interactive.key,
      }
    : undefined,
});

export const EncryptClue = (
  clue: ClueConfig,
  encrypted: boolean,
  secretKey: string,
) => ({
  /* Decoded fields */
  id: clue.id,
  /* Encoded fields */
  url: Encrypt(clue.url, encrypted, secretKey),
  text: wrapEncrypt(clue.text, encrypted, secretKey),
  html: wrapEncrypt(clue.html, encrypted, secretKey),
  image: wrapEncrypt(clue.image, encrypted, secretKey),
  alt: wrapEncrypt(clue.alt, encrypted, secretKey),
  interactive: clue.interactive
    ? {
        prompt: wrapEncrypt(clue.interactive.prompt, encrypted, secretKey),
        key: clue.interactive.key,
      }
    : undefined,
});
