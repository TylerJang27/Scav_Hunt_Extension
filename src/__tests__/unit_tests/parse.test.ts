import * as fs from "fs";
import path from "path";
import { ParseConfig } from "src/utils/parse";

const PATH_TO_TEST_DATA = "src/__tests__/unit_tests/test_data";
const PATH_TO_BAD_TEST_DATA = path.join(PATH_TO_TEST_DATA, "bad");
const PATH_TO_GOOD_TEST_DATA = path.join(PATH_TO_TEST_DATA, "good");

const failureMessages = new Map<string, string | RegExp>([
  [
    "bad_version",
    /Supplied version 0.\d. Version must be one of \[((1.\d)(, )?)+\]/,
  ],
  ["bad_index", "1"],
  ["duplicate_index", "1"],
  ["empty", "Received empty field for clues"],
  ["index_gap", "Parsed invalid index 3. Expected 2"],
  ["malformed_clue", "Missing field 'url' in clue index 1"],
  ["malformed", "Missing field 'name'"],
  ["markdown_and_text", "Exactly one of text and markdown can be set"],
  ["no_clues", "Received empty field for clues"],
  ["unordered", "Parsed invalid index 2. Expected 1"],
]);

jest.mock("src/providers/chrome");

// This test presumes that the input file is well-formed JSON but malformed JSON.
fs.readdirSync(PATH_TO_BAD_TEST_DATA).forEach((version) => {
  describe(`Parse malformed ${version} hunt config`, () => {
    const testDataVersionPath = path.join(PATH_TO_BAD_TEST_DATA, version);
    const testFiles = fs
      .readdirSync(testDataVersionPath)
      .filter((value) => value.endsWith(".json"));

    testFiles.forEach((testFile) => {
      // trunk-ignore(eslint/jest/valid-title)
      it(testFile, () => {
        const testName = path.parse(testFile).name;
        const testFilePath = path.join(testDataVersionPath, testFile);
        // trunk-ignore(eslint/@typescript-eslint/no-unsafe-assignment)
        const huntConfigInput = JSON.parse(
          fs.readFileSync(testFilePath).toString(),
        );

        const expectedMessage = failureMessages.get(testName);
        if (!expectedMessage) {
          throw new Error(`Missing expected error output for ${testName}`);
        }

        // Expected error handling case
        expect(() => ParseConfig(huntConfigInput)).toThrow(expectedMessage);
      });
    });
  });
});

fs.readdirSync(PATH_TO_GOOD_TEST_DATA).forEach((version) => {
  describe(`Parse valid ${version} hunt config`, () => {
    const testDataVersionPath = path.join(PATH_TO_GOOD_TEST_DATA, version);
    const testFiles = fs
      .readdirSync(testDataVersionPath)
      .filter((value) => value.endsWith(".json"));

    testFiles.forEach((testFile) => {
      // trunk-ignore(eslint/jest/valid-title)
      it(testFile, () => {
        const testFilePath = path.join(testDataVersionPath, testFile);
        // trunk-ignore(eslint/@typescript-eslint/no-unsafe-assignment)
        const huntConfigInput = JSON.parse(
          fs.readFileSync(testFilePath).toString(),
        );

        expect(ParseConfig(huntConfigInput)).toMatchSnapshot();
      });
    });
  });
});
