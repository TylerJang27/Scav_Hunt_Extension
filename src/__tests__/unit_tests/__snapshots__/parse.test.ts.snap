// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse valid v1.0 hunt config basic.json 1`] = `
{
  "author": "M. Test",
  "background": "graphics/background.png",
  "beginning": "This is the beginning",
  "clues": [
    {
      "alt": "Khan Academy",
      "id": 1,
      "image": "https://i.pcmag.com/imagery/reviews/07AxdIVbQ63OEkJoPgCybXt-19.1594914797.fit_scale.size_1028x578.png",
      "interactive": undefined,
      "markdown": undefined,
      "text": "The first clue",
      "url": "google.com",
    },
  ],
  "description": "Test Data",
  "encrypted": false,
  "name": "Test Hunt",
  "options": {
    "inOrder": false,
    "silent": false,
  },
  "version": "1.0",
}
`;

exports[`Parse valid v1.0 hunt config encoded.json 1`] = `
{
  "author": "M. Test",
  "background": "graphics/background.png",
  "beginning": "This is the beginning",
  "clues": [
    {
      "alt": "Sr2thvh0blifB3B1Yu25Fuk6ZcWg175",
      "id": 1,
      "image": "afHdRw0lcwH6Me6iLryn97pzLsnsByjfblW6FqnyLtm2Njv3bjSk9qphbcW5FznzZcXdJ15mL83nJdlud4mzlxlad936MovlMkDgdmB5eaGiRhJtV3m4JyR9NmjnNaPoRwWut4Kjb21uBhn1Qq3rldijW7H2QztxMyTxkcu7MkT4U15dN5D9kvx5NjDmcs5xNqyx5pmtayXaR3fxcm2eNqhfbqG1Usu4ch2zlz6uZ4Vb88xmM4DqIh4ie3DvUs3fO0Cy53wvbbmmch=",
      "interactive": undefined,
      "markdown": undefined,
      "text": "VvGohwlqI5GgZ7pccinnNi0eItGfN3s0ddWkU2=",
      "url": "Zb2z98v0Zl2gxqlfLmmiNwvwbbQw=k=",
    },
  ],
  "description": "Test Data",
  "encrypted": true,
  "name": "Test Hunt",
  "options": {
    "inOrder": false,
    "silent": false,
  },
  "version": "1.0",
}
`;

exports[`Parse valid v1.0 hunt config extra.json 1`] = `
{
  "author": "M. Test",
  "background": "graphics/background.png",
  "beginning": "This is the beginning",
  "clues": [
    {
      "alt": "Khan Academy",
      "id": 1,
      "image": "https://i.pcmag.com/imagery/reviews/07AxdIVbQ63OEkJoPgCybXt-19.1594914797.fit_scale.size_1028x578.png",
      "interactive": undefined,
      "markdown": undefined,
      "text": "The first clue",
      "url": "google.com",
    },
  ],
  "description": "Test Data",
  "encrypted": false,
  "name": "Test Hunt",
  "options": {
    "inOrder": false,
    "silent": false,
  },
  "version": "1.0",
}
`;

exports[`Parse valid v1.0 hunt config markdown.json 1`] = `
{
  "author": "M. Test",
  "background": "graphics/background.png",
  "beginning": "This is the beginning",
  "clues": [
    {
      "alt": "Khan Academy",
      "id": 1,
      "image": "https://i.pcmag.com/imagery/reviews/07AxdIVbQ63OEkJoPgCybXt-19.1594914797.fit_scale.size_1028x578.png",
      "interactive": undefined,
      "markdown": undefined,
      "text": "The first clue",
      "url": "google.com",
    },
    {
      "alt": "Khan Academy",
      "id": 2,
      "image": "https://i.pcmag.com/imagery/reviews/07AxdIVbQ63OEkJoPgCybXt-19.1594914797.fit_scale.size_1028x578.png",
      "interactive": undefined,
      "markdown": "### Subheader

The first clue",
      "text": undefined,
      "url": "google.com",
    },
  ],
  "description": "Test Data",
  "encrypted": false,
  "name": "Test Hunt",
  "options": {
    "inOrder": false,
    "silent": false,
  },
  "version": "1.0",
}
`;
